zipkin/src/main/java/com/linecorp/armeria/common/tracing/AbstractTracingBuilder.java

```diff
@@ -0,0 +1,249 @@
+     */
+    public B tracing(Tracing tracing) {
+        checkState(tracingBuilder == null, "tracingBuilder is already set.");
+        this.tracing = requireNonNull(tracing, "tracing");
+        return self();
+    }
+
+    /**
+     * Sets the lower-case label of the remote node.
+     *
+     * @throws IllegalStateException if {@link #tracing(Tracing)} is already set
+     *
+     * @see Builder#localServiceName(String)
+     */
+    public B localServiceName(String localServiceName) {
+        checkTracingAndBuilder();
+        tracingBuilder.localServiceName(localServiceName);
+        return self();
+    }
+
+    /**
+     * Sets the text representation of the primary IP address associated with this service.
+     *
+     * @throws IllegalStateException if {@link #tracing(Tracing)} is already set
+     *
+     * @see Builder#localIp(String)
+     * @see #localServiceName(String)
+     * @see #localPort(int)
+     */
+    public B localIp(String localIp) {
+        checkTracingAndBuilder();
+        tracingBuilder.localIp(localIp);
+        return self();
+    }
+
+    /**
+     * Sets the primary listen port associated with this service.
+     *
+     * @throws IllegalStateException if {@link #tracing(Tracing)} is already set
+     *
+     * @see Builder#localPort(int)
+     * @see #localIp(String)
+     */
+    public B localPort(int localPort) {
+        checkTracingAndBuilder();
+        tracingBuilder.localPort(localPort);
+        return self();
+    }
+
+    /**
+     * Sets the span {@link Reporter}.
+     *
+     * @throws IllegalStateException if {@link #tracing(Tracing)} is already set
+     *
+     * @see Builder#spanReporter(Reporter)
+     */
+    public B spanReporter(Reporter<Span> spanReporter) {
+        checkTracingAndBuilder();
+        tracingBuilder.spanReporter(spanReporter);
+        return self();
+    }
+
+    /**
+     * Sets the {@link Clock}.
+     *
+     * @throws IllegalStateException if {@link #tracing(Tracing)} is already set
+     *
+     * @see Builder#clock(Clock)
+     */
+    public B clock(Clock clock) {
+        checkTracingAndBuilder();
+        tracingBuilder.clock(clock);
+        return self();
+    }
+
+    /**
+     * Sets the {@link Sampler}.
+     *
+     * @throws IllegalStateException if {@link #tracing(Tracing)} is already set
+     *
+     * @see Builder#sampler(Sampler)
+     */
+    public B sampler(Sampler sampler) {
+        checkTracingAndBuilder();
+        tracingBuilder.sampler(sampler);
+        return self();
+    }
+
+    /**
+     * Sets the {@link CurrentTraceContext}. If this is not set and {@link #tracing(Tracing)} is not invoked,
+     * {@link RequestContextCurrentTraceContext} will be set by default.
+     *
+     * @throws IllegalStateException if {@link #tracing(Tracing)} is already set
+     *
+     * @see Builder#currentTraceContext(CurrentTraceContext)
+     */
+    public B currentTraceContext(CurrentTraceContext currentTraceContext) {
```

Comment:
Don't we need to hide this method from users because otherwise a user can specify other implementation than `RequestContextCurrentTraceContext`? We could add a delegated `addScopeDecorator()` instead.